// @nguniversal/builders/src/static-generator/index.js
"use strict";
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = void 0;
const architect_1 = require("@angular-devkit/architect");
const normalize_optimization_1 = require("@angular-devkit/build-angular/src/utils/normalize-optimization");
const service_worker_1 = require("@angular-devkit/build-angular/src/utils/service-worker");
const express_1 = __importDefault(require("express"));
const http = __importStar(require("http"));
const ora_1 = __importDefault(require("ora"));
const os_1 = require("os");
const path = __importStar(require("path"));
const piscina_1 = __importDefault(require("piscina"));
const util_1 = require("util");
const utils_1 = require("../ssr-dev-server/utils");
const utils_2 = require("./utils");
/**
 * Builds the browser and server, then renders each route in options.routes
 * and writes them to prerender/<route>/index.html for each output path in
 * the browser result.
 */
async function execute(options, context) {
    const browserTarget = (0, architect_1.targetFromTargetString)(options.browserTarget);
    const browserOptions = (await context.getTargetOptions(browserTarget));
    const routes = await (0, utils_2.getRoutes)(options, browserOptions.tsConfig, context);
    if (!routes.length) {
        throw new Error(`Could not find any routes to generate.`);
    }
    const { result } = await context.scheduleTarget(browserTarget, {
        watch: false,
        serviceWorker: false,
    });
    const { success, error, outputs } = (await result);
    if (!success) {
        return { success, error };
    }
    const { styles: normalizedStylesOptimization } = (0, normalize_optimization_1.normalizeOptimization)(browserOptions.optimization);
    const worker = createWorker();
    try {
        for (const { path: outputPath } of outputs) {
            const spinner = (0, ora_1.default)(`Prerendering ${routes.length} route(s) to ${outputPath}...`).start();
            const staticServer = await createStaticServer(outputPath);
            try {
                await Promise.all(routes.map((route) => {
                    const options = {
                        inlineCriticalCss: normalizedStylesOptimization.inlineCritical,
                        outputPath,
                        route,
                        port: staticServer.port,
                    };
                    return worker.run(options, { name: 'render' });
                }));
                spinner.succeed(`Prerendering routes to ${outputPath} complete.`);
                if (browserOptions.serviceWorker) {
                    const swResult = await generateServiceWorker(context, outputPath, browserOptions);
                    if (!swResult.success) {
                        return swResult;
                    }
                }
            }
            catch (error) {
                spinner.fail(`Prerendering routes to ${outputPath} failed.`);
                return { success: false, error: error.message };
            }
            finally {
                await staticServer.close();
            }
        }
        return { success: true };
    }
    finally {
        void worker.destroy();
    }
}
exports.execute = execute;
async function createStaticServer(browserOutputRoot) {
    const app = (0, express_1.default)();
    app.use(express_1.default.static(browserOutputRoot));
    const port = await (0, utils_1.getAvailablePort)();
    const server = new http.Server(app);
    await new Promise((res) => server.listen(port, res));
    return {
        close: (0, util_1.promisify)(server.close.bind(server)),
        port,
    };
}
function createWorker() {
    const maxThreads = Math.max(Math.min((0, os_1.cpus)().length, 6) - 1, 1);
    const worker = new piscina_1.default({
        filename: path.join(__dirname, 'worker.js'),
        name: 'render',
        maxThreads,
    });
    return worker;
}
async function generateServiceWorker(context, outputPath, browserOptions) {
    const spinner = (0, ora_1.default)(`Generating service worker for ${outputPath}...`).start();
    try {
        const projectName = context.target?.project;
        if (!projectName) {
            throw new Error('The builder requires a target.');
        }
        const projectMetadata = await context.getProjectMetadata(projectName);
        const projectRoot = path.join(context.workspaceRoot, projectMetadata.root ?? '');
        await (0, service_worker_1.augmentAppWithServiceWorker)(projectRoot, context.workspaceRoot, outputPath, browserOptions.baseHref || '/', browserOptions.ngswConfigPath);
        spinner.succeed(`Service worker generation for ${outputPath} complete.`);
        return { success: true };
    }
    catch (error) {
        spinner.fail(`Service worker generation for ${outputPath} failed.`);
        return { success: false, error: error.message };
    }
}
exports.default = (0, architect_1.createBuilder)(execute);
